### PR TITLE
[soltest] Add support for hex string literals

### PR DESCRIPTION
**Pull request by erak**
Part of #4223.

This PR aims to add support for hex string literals in (i)soltest. Hex string literals result in a left-aligned, unpadded byte representation:

```
contract C {
    function s() public returns (uint256) {
        return msg.data.length;
    }
}
// ----
// s(): hex"4200ef" -> 7
```
- [x] Change target to `develop` after https://github.com/ethereum/solidity/pull/6183 got merged

### CODE REVIEW

**Comment by erak** [comment]
Rebased.

File: test/libsolidity/semanticTests/smoke_test.sol [hunk 1]
```diff
@@ -23,3 +26,5 @@ contract C {
 // i() # Does not exist. # -> FAILURE # Reverts. #
 // x(bytes32): 0x31 -> 0x31
 // t(bool): true -> false
+// s(): hex"4200ef" -> 7
```
Comment (chriseth) [inline test/libsolidity/semanticTests/smoke_test.sol:31]: I'm not totally sure anymore we are using non-confusing styles here...

I'm pretty sure I will use things like `0x20` at some point and not remember that it will left-align.

Perhaps it would be better to still interpret hex numbers as right-aligned numbers and instead have something like

`hexLeft"4200ef"` for left-aligned padded and then `hexUnpadded"4200ef` for unpadded? Or perhaps `rawHex`?

Comment (ekpyron) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: I would say hex strings ``hex"00abcd"`` should always only be unpadded hex values of the size given by the string length - that's the only reason for using strings and consistent with solidity.

So if ``0x20`` should default to right-aligning (I can see the point for potential confusion otherwise), then I'd rather introduce ``left(0x20)`` for left-aligning them, instead of using hex strings for that.

Comment (ekpyron) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: The test case should probably have leading and trailing zeros btw.

Comment (erak) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: Yes, I agree that the syntax became a little bit confusing. I would propose to use the following:

- `0x20` (right-aligned, padded)
- `right(0x20)` (right-aligned, padded)
- `left(0x20)` (left-aligned, padded)
- `hex"4200ef"` (unpadded)

I thought about different ways of how to that and concluded a built-in function to be a proper solution. There is no real to add `right(0x20)`, but I'd do that to make concept of alignment more complete.

Comment (erak) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: I think these should not be allowed:
- `hex"0x20"`
- `left(hex"4200ef")`

Comment (chriseth) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: if that's not too hard to parse, then it looks good, I would say.

Comment (erak) [reply test/libsolidity/semanticTests/smoke_test.sol:31]: No, that shouln't be too hard. If we agree on the last proposal, we could merge this PR and I'll do the change in another one.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `9da4b63360a71491c686cda189e1c575590d4751`:
```
Error: Trailing whitespace found:
 test/libsolidity/semanticTests/smoke_test.sol:21: 

```
Please check that your changes are working as intended.

**Comment by erak** [comment]
Rebased onto https://github.com/ethereum/solidity/pull/6183 which should get merged first.

**Comment by leonardoalt** [comment]
Needs rebasing.

**Comment by erak** [comment]
Updated.

**Comment by erak** [comment]
Ready for final review. Using `left(hex"42fefe")` or `right(hex"42fefe")` are not allowed anymore.

**Comment by ekpyron** [comment]
Actually, I only now understand the question @chriseth had earlier: so far we don't have a way to encode left-aligned values with a prefix of zeros, do we?

So how would we encode
``0000123400000000000000000000000000000000000000000000000000000000``
?

``left(0x00001234)`` will result in
``1234000000000000000000000000000000000000000000000000000000000000``
won't it?

So there are two options: either we make ``left(0x00001234)`` take the leading zeros into account or we allow ``left(hex"00001234")`` after all (for which ``left`` is not a no-op, but adds padding on the right).

**Review by ekpyron — APPROVED** [review]
Despite my comment above, I'd still be fine to merge this for now and tackle the left-aligned-with-leading-zeros-case separately.

Or we change this to allow ``left(hex"00001234")`` (with padding) right away - as you like.
